bob-backend/src/pearl/group.rs

```diff
@@ -240,59 +242,69 @@ impl Group {
         if let Err(e) = result {
             // if we receive WorkDirUnavailable it's likely disk error, so we shouldn't restart one
             // holder but instead try to restart the whole disk
-            if e.is_possible_disk_disconnection() {
-                return Err(e);
-            }
-            if !e.is_duplicate() && !e.is_not_ready() {
+            if !e.is_possible_disk_disconnection() && !e.is_duplicate() && !e.is_not_ready() {
                 error!("pearl holder will restart: {:?}", e);
                 holder.try_reinit().await?;
                 holder.prepare_storage().await?;
                 debug!("backend pearl group put common storage prepared");
             }
+            Err(e)
+        } else {
+            Ok(())
         }
-        Ok(())
     }
 
     pub async fn get(&self, key: BobKey) -> Result<BobData, Error> {
         let holders = self.holders.read().await;
         let mut has_error = false;
-        let mut results = vec![];
+        let mut max_timestamp = None;
+        let mut result = None;
         for holder in holders
             .iter_possible_childs_rev(&Key::from(key))
             .map(|(_, x)| &x.data)
         {
             let get = Self::get_common(&holder, key).await;
             match get {
-                Ok(data) => {
-                    trace!("get data: {:?} from: {:?}", data, holder);
-                    results.push(data);
-                }
-                Err(err) => {
-                    if err.is_key_not_found() {
+                Ok(data) => match data {
+                    ReadResult::Found(data) => {
+                        trace!("get data: {:?} from: {:?}", data, holder);
+                        let ts = data.meta().timestamp();
+                        if ts > max_timestamp.unwrap_or(0) {
+                            max_timestamp = Some(ts);
+                            result = Some(data);
+                        }
+                    }
+                    ReadResult::Deleted(ts) => {
+                        trace!("{} is deleted in {:?} at {}", key, holder, ts);
+                        let ts: u64 = ts.into();
+                        if ts >= max_timestamp.unwrap_or(0) {
```

Comment:
Possible problem: `ts` from `ReadResult::Deleted(ts)` is a Pearl level timestamp, while `data.meta().timestamp()` is a user level timestamp. They are different and should not be mixed. Proper way is to compare only user level timestamps!
Please, add timestamp as a parameter of Delete request on gRPC protocol level.

Reply:
https://github.com/qoollo/bob/issues/708